services/auth/httpsign.go

```diff
@@ -0,0 +1,183 @@
+	// HTTPSign authentication should only fire on API
+	if !middleware.IsAPIPath(req) {
+		return nil
+	}
+
+	sigHead := req.Header.Get("Signature")
+	if len(sigHead) == 0 {
+		return nil
+	}
+
+	if len(setting.SSH.TrustedUserCAKeys) == 0 {
+		return nil
+	}
+
+	validpk, err := VerifyCert(req)
+	if err != nil {
+		log.Warn("Failed authentication attempt from %s", req.RemoteAddr)
+		log.Error("VerifyCert failed: %v", err)
+		return nil
+	}
+
+	u, err := user_model.GetUserByID(validpk.OwnerID)
+	if err != nil {
+		log.Error("GetUserByID:  %v", err)
+		return nil
+	}
+
+	store.GetData()["IsApiToken"] = true
+
+	log.Trace("HTTP Sign: Logged in user %-v", u)
+
+	return u
+}
+
+// VerifyCert verifies the validity of the ssh certificate and returns the publickey of the signer
+// We verify that the certificate is signed with the correct CA
+// We verify that the http request is signed with the private key (of the public key mentioned in the certificate)
+func VerifyCert(r *http.Request) (*asymkey_model.PublicKey, error) {
+	var validpk *asymkey_model.PublicKey
+
+	// Get our certificate from the header
+	bcert, err := base64.RawStdEncoding.DecodeString(r.Header.Get("x-ssh-certificate"))
+	if err != nil {
+		return validpk, err
+	}
+
+	pk, err := ssh.ParsePublicKey(bcert)
+	if err != nil {
+		return validpk, err
+	}
+
+	// Check if it's really a ssh certificate
+	cert, ok := pk.(*ssh.Certificate)
+	if !ok {
+		return validpk, fmt.Errorf("no certificate found")
+	}
+
+	for _, principal := range cert.ValidPrincipals {
+		validpk, err = asymkey_model.SearchPublicKeyByContentExact(r.Context(), principal)
+		if err != nil {
+			if asymkey_model.IsErrKeyNotExist(err) {
+				continue
+			}
+			log.Error("SearchPublicKeyByContentExact: %v", err)
+			return validpk, err
+		}
+
+		c := &ssh.CertChecker{
+			IsUserAuthority: func(auth ssh.PublicKey) bool {
+				for _, k := range setting.SSH.TrustedUserCAKeysParsed {
+					if bytes.Equal(auth.Marshal(), k.Marshal()) {
+						return true
+					}
+				}
+
+				return false
+			},
+		}
+
+		// check the CA of the cert
+		if !c.IsUserAuthority(cert.SignatureKey) {
+			return validpk, fmt.Errorf("CA check failed")
+		}
+
+		// validate the cert for this principal
+		if err := c.CheckCert(principal, cert); err != nil {
+			return validpk, fmt.Errorf("no valid principal found")
+		}
+
+		break
+	}
+
+	// validpk will be nil when we didn't find a principal matching the certificate registered in gitea
+	if validpk == nil {
+		return validpk, fmt.Errorf("no valid principal found")
+	}
+
+	verifier, err := httpsig.NewVerifier(r)
+	if err != nil {
+		return validpk, fmt.Errorf("httpsig.NewVerifier failed: %s", err)
+	}
+
+	// now verify that we signed this request with the publickey of the cert
+	err = doVerify(verifier, []ssh.PublicKey{cert.Key})
+	if err != nil {
+		return validpk, err
+	}
+
+	return validpk, nil
+}
+
+func doVerify(verifier httpsig.Verifier, publickeys []ssh.PublicKey) error {
```

Comment:
Needs a comment.

Perhaps something like below?

```suggestion
// doVerify iterates across the provided public keys attempting the verify the current request against each key in turn
func doVerify(verifier httpsig.Verifier, publickeys []ssh.PublicKey) error {
```